core/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import org.elasticsearch.common.collect.CopyOnWriteHashMap;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.index.mapper.object.ObjectMapper;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * An immutable container for looking up {@link MappedFieldType}s by their name.
+ */
+class FieldTypeLookup implements Iterable<MappedFieldType> {
+    private static final Function<MappedFieldTypeReference, MappedFieldType> UNWRAPPER = new Function<MappedFieldTypeReference, MappedFieldType>() {
+        @Override
+        public MappedFieldType apply(MappedFieldTypeReference ref) {
+            return ref.get();
+        }
+    };
+
+    /** Full field name to field type */
+    private final CopyOnWriteHashMap<String, MappedFieldTypeReference> fullNameToFieldType;
+
+    /** Index field name to field type */
+    private final CopyOnWriteHashMap<String, MappedFieldTypeReference> indexNameToFieldType;
+
+    /** Create a new empty instance. */
+    public FieldTypeLookup() {
+        fullNameToFieldType = new CopyOnWriteHashMap<>();
+        indexNameToFieldType = new CopyOnWriteHashMap<>();
+    }
+
+    private FieldTypeLookup(CopyOnWriteHashMap<String, MappedFieldTypeReference> fullName, CopyOnWriteHashMap<String, MappedFieldTypeReference> indexName) {
+        fullNameToFieldType = fullName;
+        indexNameToFieldType = indexName;
+    }
+
+    /**
+     * Return a new instance that contains the union of this instance and the field types
+     * from the provided fields. If a field already exists, the field type will be updated
+     * to use the new mappers field type.
+     */
+    public FieldTypeLookup copyAndAddAll(Collection<FieldMapper> newFieldMappers) {
```

Comment:
No, this is not called concurrently (protected by the mapping write lock in MapperService/DocumentMappers). I dont think the COW map will allow inconsistent views, since an entire new lookup is created (so any new fields will not be visible to search until this method returns and MapperService updates its lookup reference). We do have consistency problems, though. For example. DocumentMapper has its own map of fields (FieldMapers, not MappedFieldTypes) which is updated first, and then MapperService is updated. But these problems all already existed (this lookup class was formerly called FieldMappersLookup, I just renamed it when the type changed).  I'm happy to continue working towards full immutability in the mappings, but I think that should happen in follow up issues.